GeneticAlgorithm.java

```diff
@@ -84,89 +87,86 @@ private static double evaluateIndividual (Genome individual, State s) {
-        currBestScore = 0;
-        int saveId = bestId;
-        evaluationScore.put(saveId, Double.NEGATIVE_INFINITY);
-        for (Integer id : evaluationScore.keySet()) {
-            if(evaluationScore.get(id) > currBestScore){
-                bestId = id;
-                currBestScore = evaluationScore.get(id);
+        numParents--;
+        for (Integer id : sortedScore.keySet()) {
+            if(numParents >= 0) {
+                parents[numParents] = id;
+            } else {
+                break;
             }
+            numParents--;
         }
-        parents[1] = bestId;
-        evaluationScore.put(saveId, bestScore);
 
         return parents;
     }
 
-    /**
-     * return the ids of parents
-     */
-    public static int selectReplacement(HashMap<Integer, Double> evaluationScore) {
-        int replaceId = 0;
-        double currScore = Double.POSITIVE_INFINITY;
-
-        for (Integer id : evaluationScore.keySet()) {
-            if(evaluationScore.get(id) < currScore){
-                replaceId = id;
-                currScore = evaluationScore.get(id);
-            }
+    public static ArrayList<Genome> updatePopulation(int[] parents, ArrayList<Genome> population) {
+        int childrenNum = population.size() - parents.length;
+        ArrayList<Genome> newPopulation = new ArrayList<Genome>();
+
+        //add parents in newPopulation
+        for (int i = 0; i< parents.length; i++) {
+            newPopulation.add(population.get(parents[i]));
+        }
+
+        for (int i = 0; i < childrenNum; i++) {
+            int randomIndexMom = (int) Math.floor(Math.random() * parents.length);
+            int randomIndexDad = (int) Math.floor(Math.random() * parents.length);
+
+            Genome momGenome = population.get(randomIndexMom);
+            Genome dadGenome = population.get(randomIndexDad);
+
+            /* cross over */
+            Random random = new Random();
+            double holesWeight = random.nextBoolean() ? momGenome.getHolesWeight() : dadGenome.getHolesWeight();
+            double heightDifferenceWeight = random.nextBoolean() ? momGenome.getHeightDifferenceWeight() : dadGenome.getHeightDifferenceWeight();
+            double islandWeight = random.nextBoolean() ? momGenome.getIslandWeight() : dadGenome.getIslandWeight();
+            double parityWeight = random.nextBoolean() ? momGenome.getParityWeight() : dadGenome.getParityWeight();
+            double edgeHeightWeight = random.nextBoolean() ? momGenome.getEdgeHeightWeight() : dadGenome.getEdgeHeightWeight();
+            double blockadeWeight = random.nextBoolean() ? momGenome.getBlockadeWeight() : dadGenome.getBlockadeWeight();
+            Genome child = new Genome(newPopulation.size()+1, blockadeWeight, edgeHeightWeight, heightDifferenceWeight, holesWeight, islandWeight, parityWeight);
+
+            /* mutation */
+            int randomIndex = (int) Math.floor(Math.random() * (numberOfFeatures - 1));
```

Comment:
I think for mutation, we shouldn't pick one index to mutate. The conventional way is like this:
```
for each gene:
  if random() > p:
    mutate()
```
where `p` is a very small value. This way, it's possible for no mutation to happen (to retain optimal gene combinations). 

(https://en.wikipedia.org/wiki/Mutation_(genetic_algorithm))